src/lkl/setup.c

```diff
@@ -402,28 +384,43 @@ static void* lkl_activate_crypto_disk_thread(struct lkl_crypt_device* lkl_cd)
 
     crypt_free(cd);
 
-    // The key is only needed during activation, so don't keep it around
-    // afterwards and free up space.
+    // The key is only needed during activation, so clear it now.
     memset(lkl_cd->disk_config->key, 0, lkl_cd->disk_config->key_len);
-
-    unsigned long munmap_ret;
-    if ((munmap_ret = lkl_sys_munmap(
-             (unsigned long)lkl_cd->disk_config->key,
-             lkl_cd->disk_config->key_len)))
-    {
-        sgxlkl_fail(
-            "Unable to unmap memory for disk encryption key: %s\n",
-            lkl_strerror((int)munmap_ret));
-    }
     lkl_cd->disk_config->key = NULL;
     lkl_cd->disk_config->key_len = 0;
 
     return 0;
 }
 
+// ATTN: integrity is disabled for now because dm-ioctl() depletes kernel
+// memory when formatting large integrity devices (when executing within the
+// dm-integrity module). This can be overcome either by making the disk image
+// small or making the kernel memory large. The following turns out to work
+// for 3GB disks.
+//
+//     export SGXLKL_CMDLINE="mem=80M"
+//
+// But allocating this much memory to the kernel is unreasonable. The only
+// workable solution seems to be formatting integrity devices in user space.
+// It is unclear why dm-crypt and dm-verity formatting is done in user space
+// while dm-intetrity formatting is done in kernel space.
+//
+// The failure begins in vic_dm_create_integrity() and occcurs when libdevmapper
+// invokes icotl() with the DM_RELOAD_CMD request. The kernel then panics with
+// an out-of-memory error. Custom device-mapper user-space code was written
+// that bypasses libdevmapper but the error is the same.
+//
+// Recall that integrity formatting is performed by the dm-integrity module
+// within the kernel. The integrity superblock is initially zero-filled. When
+// activated for the first time, the kernel module formats the device.
+// #define ENABLE_INTEGRITY
+
 static void* lkl_create_crypto_disk_thread(struct lkl_crypt_device* lkl_cd)
 {
     int err;
+    /* ATTN: vicsetup only supports 512 bytes sectors for integrity */
+    static const size_t SECTOR_SIZE = 512;
+    static const size_t MIN_LUKS2_ITERATIONS = 1000;
```

Comment:
A comment explaining what this sets would be useful.

Reply:
Done. Thank you.